INT-3846: Some fixes for the `SimpleMessageStore`

artembilan opened the pull request:
JIRAs:
https://jira.spring.io/browse/INT-3830
https://jira.spring.io/browse/INT-3523
https://jira.spring.io/browse/INT-3846
- Fix the OOM condition, when we `release()` `UpperBound` independently of the previous `remove` result (https://jira.spring.io/browse/INT-3846)
- Fix "confuse" around `groupCapacity`, when we really didn't care about individual groups (https://jira.spring.io/browse/INT-3523)
- Add `upperBoundTimeout` to have a hook to wait some time for the empty slot in the store (https://jira.spring.io/browse/INT-3830)
- Fix some JavaDocs warnings
- Fix some typos

Commit:
INT-3846: Some fixes for the `SimpleMessageStore`

JIRAs:
https://jira.spring.io/browse/INT-3830
https://jira.spring.io/browse/INT-3523
https://jira.spring.io/browse/INT-3846

* Fix the OOM condition, when we `release()` `UpperBound` independently of the previous `remove` result (https://jira.spring.io/browse/INT-3846)
* Fix "confuse" around `groupCapacity`, when we really didn't care about individual groups (https://jira.spring.io/browse/INT-3523)
* Add `upperBoundTimeout` to have a hook to wait some time for the empty slot in the store (https://jira.spring.io/browse/INT-3830)
* Fix some JavaDocs warnings
* Fix some typos

artembilan commented:
@celandro, 
this is a fresh PR, which includes your fixes and my polishing.

Feel free to comment!

Thank you very much for the contribution any way!

celandro commented:
Looks great! I really like the per group limit, quite clean

celandro commented on spring-integration-core/src/main/java/org/springframework/integration/store/SimpleMessageStore.java:
```diff
@@ -233,8 +270,8 @@ public void removeMessageGroup(Object groupId) {
 
-				groupUpperBound.release(groupIdToMessageGroup.get(groupId).size());
-				groupIdToMessageGroup.remove(groupId);
+				this.groupToUpperBound.remove(groupId).release(this.groupCapacity);
```
Can this NPE if groupId does not exist?

artembilan replied:
I don't think so. Because the code here is enough robust from the concurrent access:
```
lock.lockInterruptibly();
try {
        if (!groupIdToMessageGroup.containsKey(groupId)) {
	    return;
	}
	this.groupToUpperBound.remove(groupId).release(this.groupCapacity);
	this.groupIdToMessageGroup.remove(groupId);
}
finally {
	lock.unlock();
}
```
You can see the `lock` and a couple of modification operators if `groupId` exists.

celandro replied:
Agree, i checked the addMessageToGroup code to make sure. Id still be paranoid about it and add a null check to future proof the code on  this.groupToUpperBound.remove(groupId)

garyrussell replied:
+1

artembilan replied:
I've revised this and it does not make sense for me to `release` that `UpperBound` any more.
If if remove the `group`, we should just clean everything around it.
If something is waiting for the permits (]addMessageToGroup]), it should create a new `UpperBound`.

celandro commented:
If you can get this build on maven central, I can load test it before your release this week. Its a pain to test a foreign snapshot in our build system or I'd test it off this branch.

artembilan commented:
@celandro ,
thank you for reaching this PR and your review!

We don't place any SNAPSHOT, or even M or RC release to the Maven Central, but you can test it against Spring's repository:
```
repositories {
	maven { url 'http://repo.spring.io/libs-snapshot' }
}
```
We let you know when it will be available there.

BTW, this PR is only for `master` - `4.2.1` version.
For backporting I'll raise different PRs, but only after acceptance of this one.

CC @garyrussell

celandro commented:
Found a major bug in DelayHandler
```java
	private boolean removeDelayedMessageFromMessageStore(Message<?> message) {
		if (this.messageStore instanceof SimpleMessageStore) {
			SimpleMessageGroup messageGroup =
					(SimpleMessageGroup) this.messageStore.getMessageGroup(this.messageGroupId);
			return messageGroup.remove(message);
		}
		else {
			return ((MessageStore) this.messageStore).removeMessage(message.getHeaders().getId()) != null;
		}
	}
```

celandro commented:
This fixes the issue but I don't like it. The upperbound release has to be moved outside of the lock.
```java
	@Override
	public void removeMessagesFromGroup(Object groupId, Collection<Message<?>> messages) {
        UpperBound upperBound = this.groupToUpperBound.get(groupId);
        upperBound.release(messages.size());
		Lock lock = this.lockRegistry.obtain(groupId);
		try {
			lock.lockInterruptibly();
			try {
				SimpleMessageGroup group = this.groupIdToMessageGroup.get(groupId);
				Assert.notNull(group, "MessageGroup for groupId '" + groupId + "' " +
						"can not be located while attempting to remove Message(s) from the MessageGroup");
				boolean modified = false;
				for (Message<?> messageToRemove : messages) {
					if (group.remove(messageToRemove)) {
						modified = true;
					}
				}
				if (modified) {
					group.setLastModified(System.currentTimeMillis());
				}
			}
			finally {
				lock.unlock();
			}
		}
		catch (InterruptedException e) {
			Thread.currentThread().interrupt();
			throw new MessagingException("Interrupted while obtaining lock", e);
		}
	}
```
I'm sure there is a better way but I'm at a loss.

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/store/SimpleMessageStore.java:
```diff
@@ -208,6 +239,12 @@ public MessageGroup addMessageToGroup(Object groupId, Message<?> message) {
 					this.groupIdToMessageGroup.putIfAbsent(groupId, group);
+					this.groupToUpperBound.putIfAbsent(groupId, new UpperBound(this.groupCapacity));
+				}
+				if (!this.groupToUpperBound.get(groupId).tryAcquire(this.upperBoundTimeout)) {
```
We either need a new catch for `InterruptedException` here, or at least we need to fix the message in the `MessagingException` in the outer catch since the interrupt might be here not on the lock acquisition.

artembilan replied:
@garyrussell , unfortunately  we don't have here `InterruptedException`.
See `UpperBound.tryAcquire()` source code:
```
catch (InterruptedException e) {
	Thread.currentThread().interrupt();
	return false;
}
```

Or am I missing anything here?

garyrussell replied:
Ah, of course; yes, no problem; ignore my comment.

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/store/SimpleMessageStore.java:
```diff
@@ -255,8 +292,10 @@ public MessageGroup removeMessageFromGroup(Object groupId, Message<?> messageToR
-				group.remove(messageToRemove);
-				group.setLastModified(System.currentTimeMillis());
+				if (group.remove(messageToRemove)) {
+					this.groupToUpperBound.get(groupId).release();
```
Perhaps add a utility method for all these releases, and add an `Assert.state()` that the UB is not null.

artembilan commented:
Hello, @celandro !

Thank you very much for pointing that out!
Really good catch and it looks like nasty bug :smile: 

Well, looking to the `DelayHandler` code and this our new stuff around `UpperBound`, I'd say that we introduced a new serious **dead lock** issue.

See:
1. We call `addMessageToGroup` obtaining the `lock` for the group.
2. Reaching `this.groupToUpperBound.get(groupId).tryAcquire(this.upperBoundTimeout)` and wait for the gap in the `UpperBound`
3. We call `removeMessageFromGroup` to free a slot in the `UpperBound`.
4. And ... we can't do that, because we are waiting for `lock` here !

Not sure yet how to fix or even overcome, but I see that your concern for the `DelayHandler` should be fixed exactly from there, not `MS` part. We just must not remove from the `group` outside of `MS`.

I guess we can disable `messageGroup.remove(message)` if we are together with `SimpleMessageStore`. In other words don't allow any modifications outside of `UpperBound` for the group.

celandro commented:
That fixes 1 issue, but the issue of the tryAquire inside the lockExclusively means that you can easily exceed the maxAquire time. Moving the tryAquires in all the addMessage methods outside the lock would fix that

artembilan commented:
That is an interesting point. But how about do not rely on the `tryAquires` for the groups, but switch to the `lock.newCondition()` and free the lock for the `removeMessageFromGroup`?

Having these all concerns, I'd say that we can't bakcport it to all those `maint` versions.
Or, let's make the `upperBoundTimeout` only as a `master` feature!
We can backport those `if (message != null)` around `release` and try to fix `DelayHandler`.

But I definetelly sure all this changes and fixes are very dangerous for those versions which are here for a while already...

celandro commented:
I will make a few changes and make a new pull request after I loadtest a few million messages through a delayhandler. The code will be setup to log warnings if remove is called on messages that don't exist.

Remember that the code in the release version will hang if you ever insert maxCapacity elements into the store. An imperfect fix is still much better than the release version

artembilan commented:
> Remember that the code in the release version will hang if you ever insert maxCapacity elements into the store. An imperfect fix is still much better than the release version 

Not sure what you mean. Please, find the JavaDocs for the `Semaphore.tryAcquire`:
```
     * is returned.  If the time is less than or equal to zero, the method
     * will not wait at all.
```

As I said: this " imperfect fix" produces **dead lock**. Even if it is for some time (`upperBoundTimeout`), we don't have a gain of it. Just because we don't release any permit, when we would like.

Having that I even don't see reason to have this `UpperBound` at all. Simple counter (`AtomicInteger`) together with `Lock` and the `Condition` on it does the stuff.

celandro commented:
You are correct, it does not hang on the producer end but it prevents any new elements from being added, ever. The effect is rejecting all new entries into the store or completely breaking a delayer. My consumers of the queue look hung in this situation but its actually the message store.

As for semaphores or atomic integers, I have no opinion, I was trying for the most minimally invasive changes originally but I can certainly make this version have no race conditions and only have issues with more elements in the store than is allowed when remove is called on elements that do not exist (same issue as the current version).

I'm also a little concerned there may have been a performance degredation at some point that makes a delayer working off a large message store(~150K backlog) perform poorly. The remove methods being used are O(N) worst case, I will look into it.  Im only getting about 70/sec in my load test environment with this version, my prod version based on 2.2.6 gets 1200/sec. There may be some other issue in my environment, or the locking differences may be the culprit, Will look today.

Perhaps I should extend SimpleMessageStore instead of modifying it? or make an interface they both conform to so DelayerHandler and others can check for that?

artembilan commented:
You definitely can make your own MessageGroupStore and inject it to the DelayHandler. For this one I must think more. Come back to after the lunch

artembilan committed:
* Fix inconsistency in the `DelayHandler` around `removeMessageFromGroup` when `MS` is `SimpleMessageStore`
 * Remove `UpperBound.release()` operation from `SimpleMessageStore.removeGroup()`.
 The waiting process should worry about the new `UpperBound` instance.

artembilan commented:
Pushed some fix.
It isn't finished yet, because I'm switching computers. 
But be sure to review it as you wish!

artembilan committed:
Some other polishing

artembilan commented:
Good morning!

Pushed some simple polishing, but it isn't the final solution yet.
So, it isn't ready for the merge.

I can't still figure out the fix for dead lock.

Need to think more.
And meanwhile I go to fix other JIRAs for the tomorrow's release.

Feel free to share you thought here how to overcome that dead lock.

Thanks in advance!

artembilan committed:
`tryAcquire` outside of the `lock`

artembilan committed:
Move `tryAcquire` within the `addMessageToGroup` outside of `lock`.
But do that only for groups which already exist.
For the new groups we have a fresh `UpperBound`, so no need to worry about dead lock and
 we can obtain  a permit immediately.

artembilan commented:
Pushed.
From my perspective it is the best solution right now.
Please, review ASAP, since we would like to release all the latest fixes for all version we support.

Thanks for feedback, contribution, testing and just for the keeping dialog.

Looking forward for more contribution!

artembilan commented:
Hi @celandro !

I know you are busy, too.
Anyway would you mind to find some minute to review this one more time for the final solution.

We have some urgent plan to release :smile: 

Thank you for the contribution one more time!

celandro commented:
It looks good to me but I haven't had a chance to load test unfortunately.

celandro commented:
Unfortunately I was not able to test with this code change however I tested with just a modified extension of SimpleMessageStore.

2 issues
1) if you call remove twice with the same message (since I have not changed the delayer), the second remove will scan the entire queue which is horrendously slow. i hope this is not the case with the new code but I can not test today
2) if you remove somewhere else and then remove, it does not decrement the tryAquire. I worked around that since its not realistic

```java
    @Override
    public void removeMessagesFromGroup(Object groupId, Collection<Message<?>> messages) {
        Lock lock = this.lockRegistry.obtain(groupId);
        try {
            lock.lockInterruptibly();
            try {
                SimpleMessageGroup group = this.groupIdToMessageGroup.get(groupId);
                Assert.notNull(group, "MessageGroup for groupId '" + groupId + "' " +
                        "can not be located while attempting to remove Message(s) from the MessageGroup");
                UpperBound upperBound = this.groupToUpperBound.get(groupId);
                Assert.state(upperBound != null, "'upperBound' must not be null.");
                boolean modified = false;
                upperBound.release(messages.size());
                for (Message<?> messageToRemove : messages) {
                    if (group.remove(messageToRemove)) {
                        modified = true;
                    } else {
                        log.warn("Could not remove message, capacity may be exceeded");
                    }
                }
                if (modified) {
                    group.setLastModified(System.currentTimeMillis());
                }
            }
            finally {
                lock.unlock();
            }
        }
        catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new MessagingException("Interrupted while obtaining lock", e);
        }
    }
```

artembilan commented:
Hello, @celandro !

Thank you again for feedback!

So, let me try to answer to your concerns:
> 1) if you call remove twice with the same message (since I have not changed the delayer), the second remove will scan the entire queue which is horrendously slow. i hope this is not the case with the new code but I can not test today

Not a case for now. This my `DelayHandler` version has these code changes:
````java
private boolean removeDelayedMessageFromMessageStore(Message<?> message) {
		if (this.messageStore instanceof SimpleMessageStore) {
			synchronized (this.messageGroupId) {
				Collection<Message<?>> messages = this.messageStore.getMessageGroup(this.messageGroupId).getMessages();
				if (messages.contains(message)) {
					this.messageStore.removeMessageFromGroup(this.messageGroupId, message);
````
So, we call here `removeMessageFromGroup` just for single `Message`. It isn't an issue here because we deal with `in-memory` store. Returning the `MessageGroup` from there doesn't impact the whore performance.

Another change looks like:
````java
if (!(this.messageStore instanceof SimpleMessageStore)) {
	this.messageStore.removeMessagesFromGroup(this.messageGroupId, message);
}
````
So, we don't perform `removeMessagesFromGroup` (iterable one) for the `SimpleMessageStore`.

Therefore I don't see your issue right now.

>  if you remove somewhere else and then remove, it does not decrement the tryAquire. I worked around that since its not realistic

It isn't clear for me what is `somewhere else` from your perspective, but if you mean that my issue before, like this:
````java
messageGroup.remove(message);
````
I agree and from the `SimpleMessageStore` we should disallow any `MessageGroup` modification.
But we can't do this breaking change for the `maint` releases. It maybe like `robustness improvement` for the next release.
Right now it is up to end-user to destroy his application by his conscience.

celandro commented:
Agree remove is not an issue however .contains might be.  I would do the following:
```java
	private void doReleaseMessage(Message<?> message) {
		if (removeDelayedMessageFromMessageStore(message)) {
			this.messageStore.removeMessagesFromGroup(this.messageGroupId, message);
			this.handleMessageInternal(message);
		}
		else {
			if (logger.isDebugEnabled()) {
				logger.debug("No message in the Message Store to release: " + message +
						". Likely another instance has already released it.");
			}
		}
	}

	private boolean removeDelayedMessageFromMessageStore(Message<?> message) {
		if (this.messageStore instanceof SimpleMessageStore) {
			// not handled in here. Due to potential locking on .contains, just return true
			return true;
		}
		else {
			return ((MessageStore) this.messageStore).removeMessage(message.getHeaders().getId()) != null;
		}
	}
```
Unfortunately the contract of SimpleMessageStore is not a great fit for use in a delay store as there is no way to get the number of messages removed and it does not match the implementation of any other setup. Passing around the message itself instead of just the id and group seems strange to me as well.  I also am concerned with the contract of MessageStore in general as calling removeMessage followed by removeMessageFromGroup is unintuitive and implies that the caller needs to clean up what amounts to an index.

One more major issue is  when configurable delays are enabled, SimpleMessageGroup reverts to O(n)  performance since there is no guaranteed ordering of the .remove methods when they are called. For example, in my situation, certain events have no delay and some events have 30s delay. This causes the no delay events to scan the entire queue before removal.  These events then have the potential of blocking up the entire processing pipeline. 

I think making a version of SimpleMessageGroup that works with only ConcurrentHashMaps would be better performance. I will work on it.

artembilan commented:
>  I would do the following:

No, It's bad again. You don't check if you removed message from the store really. And therefore will end up with the issue for which that change in the `DelayHandler` exists there: do not duplicate message after releasing. 

I don't think that my `synchronized (this.messageGroupId)` around `if (messages.contains(message)) {` isn't so bad there, because we are single-threaded in the `this.messageStore.removeMessageFromGroup` anyway according to the `lock` for the `groupId`.

Regarding all your other concerns...

From one side I agree that `MessageStore` and `MessageGroupStore` contracts are a bit distructive, but we can do nothing right now on the matter. And live with them as is just because there is no premise to use them out side of Framework. So, the question to change their contract is something for the far future...

For you idea around `ConcurrentHashMap`, please, take a look here https://gist.github.com/artembilan/a7adca1cdb9c12a5d3a8. We can even consider this impl for as an out-of-the box, but already in the next release.

You last issue isn't clear to me at all...
The events without delay doesn't cause any scheduling at all:
````java
if (delay > 0) {
	this.releaseMessageAfterDelay(requestMessage, delay);
	return null;
}
````

Please, be more specific, but already in the new JIRA and with some test-case to reproduce or, at least, to play.

Let's finish with this PR and summarize if it is sufficient to merge and backport!

Actually we with @garyrussell just wait here for your acceptance :smile:

celandro commented:
I understand your desire not to touch the contract.

Looked at the version using ConcurrentHashMap and I think it looks good. I implemented something very similar as a different class since there might be some use case where they want to use it as an actual queue. Maybe name the class RandomAccessMessageStore. Regardless, I will test your version and let you know how it performs.

Sorry we had just increased the default delay to 100ms with the others waiting 30000ms, my mistake. The random access should resolve the issue anyhow.

artembilan commented:
Great! Thank you!
But again: we are talking here about the `SimpleMessageStore` bugs, but not the `DelayHandler` performance. It is definitely the different story and must be reviewed in the separate thread.
Yes, we should keep in mind any component making such a critical changes, but I feel we can sever that `DelayHandler` issue right now.

celandro commented:
I think it is good to release as well and resolves the issue with this bug. I see no reason to hold up the release. 

I will let you know about the other class you made. The only issue I see so far is the wiring is a bit awkward as you have to set the groupId to something like "delayerId.messageGroupId" and I'm about to run the load test...

celandro commented:
The new store is significantly faster than the old one and I no longer see any threads stuck in the message store.

artembilan commented:
@celandro ,

If you mean that my workaround for you, so don't count it, please, for the PR. That is just a quick fix for your particular use-case.

Our real concern here is about the purposed fix to be applied to the Framework.
So, let us know what you feel about that.

Thank you again for your valuable feedback!

celandro commented:
I'm not counting on it, I have copied into my own code and everything is working well.

This fix looks fine excepting the performance issue.

artembilan commented:
@garyrussell , would be good if you'd say your final point to this PR.
And let me know if I can go ahead with the backport one.

Thanks in advance!

celandro commented:
FYI the code on our end is in production right now which supports both the SimpleMessageStore and SimpleGroupMessageStore. Both passed our regression tests. SimpleMessageStore failed our load tests when ~100,000 messages are in the store. SimpleGroupMessageStore passed. We are configured to use SimpleGroupMessageStore right now in production. 

Thank you so much for all your work and effort on this ticket. Its a big deal for us here at Eharmony.

If you want to make a separate ticket for SimpleGroupMessageStore, when you are ready to merge that in, we would be happy to run our regression and load test suite against it.

artembilan commented:
That's good feedback, @celandro !

Can you elaborate more for this?
> SimpleMessageStore failed our load tests when ~100,000 messages are in the store. 

And I'd say that it is up to use to raise a ticket for the `SimpleGroupMessageStore`. With that you can track it. And even contribute! :smile:

celandro commented:
We could not go past ~30 requests a second. Threads were stuck in the message store. With the hashmap we were limited by other parts of the pipeline but went over 700/s with no threads stuck. The slowness was attributed to scanning the queue with variable delays with a capacity.

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/handler/DelayHandler.java:
```diff
@@ -365,9 +367,16 @@ private void doReleaseMessage(Message<?> message) {
-			return messageGroup.remove(message);
+			synchronized (this.messageGroupId) {
+				Collection<Message<?>> messages = this.messageStore.getMessageGroup(this.messageGroupId).getMessages();
+				if (messages.contains(message)) {
```
This is not so good - messages is an unmodified collection wrapping a `LinkedBlockingQueue`. `contains` has to iterate over the contents which is not good for a large group.

We had a similar problem in the `AcceptOnceFileListFilter` and we solved that (INT-3572) by maintaining a `Set` parallel to the queue; we can't use that technique here, though.

Looking at `SimpleMessageGroup` - it's really not so good for large groups - `remove()` requires iterating over the entire queue as well.

This probably explains Ryan's performance observation.

We should look at using a (synchronized) `LinkedHashSet` instead of the `BlockingQueue` and compare the performance for both small and large groups.

If using a synchronized set is too expensive, we might have to end up with 2 stores - one for small groups and one for large.

artembilan replied:
I wonder why do we use the `LinkedBlockingQueue` here at all...

garyrussell replied:
Looks like it was intentional to support concurrency....

https://github.com/spring-projects/spring-integration/commit/7bcadf6e9025d11b9d59ea22bdfd289e286920be

...poor decision.

celandro commented:
@garyrussell  please look at https://gist.github.com/artembilan/a7adca1cdb9c12a5d3a8 which @artembilan created. This is what we are using in production and does not have the performance issues. We have all independently agreed that using a queue is a poor choice :).
The scheduler inside the delayer already performs the queueing neccessary.
It might also be a good idea to switch the delayer to only keeping the id in the queue  so you don't have to keep all the messages in memory. Then any map implementation or key value store is a natural fit.

artembilan committed:
SimpleMessageGroup: BlockingQueue -> LinkedHashSet

artembilan commented:
Guys, take a look, please to the last commit and let me know if that is what you are expecting.
There are a lot of failed test with the `ConcurrentModificationException` to fix.
But, first of all, I'd like to know if I am on the right way.

Thank you!

garyrussell commented:
Yes; that's what I was thinking, but we need to be sure the synchronization is not prohibitively expensive for small groups. If it is, we might need to offer two `SimpleMessageGroup` versions.

artembilan commented:
Well...
Let's think if we need a synchronization in the `SimpleMessageGroup` because concurrency is guarantied by the `lock` in the `MessageGroupStore`.

garyrussell commented:
True (at least in our code), but we don't have any guarantees that users are not using this class outside of the MGS.

We could add an unsynchronized version as a private class in SMGS.

artembilan commented:
M-m-m. To be more safe, I suggests do not touch `SimpleMessageGroup` at all, and really return some unsynchornized from the `SMGS`. 
And from my perspective it should be `unmodifable`.
But that is fully different story which isn't related to this issue...

garyrussell commented:
The problem is the SMGS uses `group = new SimpleMessageGroup(groupId);`.

It's tightly coupled.

artembilan committed:
`SimpleMessageStore`: use "unsynchonized" `SimpleMessageGroup`

Make some synchronization fixes according to the migration to the `LinkedHashSet`

artembilan commented:
Please, find the last commit regarding our discussion about "unsynchronized" `SimpleMessageGroup`

artembilan committed:
Avoid extra `Collection`

artembilan commented:
Please, find what I did right now.
If you imagine some other solution, I'm open for the mind storm.

artembilan committed:
`ResequencingMessageHandler`: compare `size()` of collections instead of `containsAll()`

artembilan commented:
Pushed according the latest comments.

garyrussell commented:
LGTM - running a testAll because of the travis failure; will merge if all is well.

artembilan commented:
Actually, that is correct failure on Travis.
I haven't tested the whole project before pushing.
So, we still have `ConcurrentModificationException` there...

artembilan committed:
Fix `ConcurrentModificationException` in the `AbstractKeyValueMessageStore`

artembilan commented:
Pushed fix.

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/handler/DelayHandler.java:
```diff
@@ -365,9 +367,16 @@ private void doReleaseMessage(Message<?> message) {
+			synchronized (this.messageGroupId) {
+				Collection<Message<?>> messages = this.messageStore.getMessageGroup(this.messageGroupId).getMessages();
+				if (messages.contains(message)) {
+					this.messageStore.removeMessageFromGroup(this.messageGroupId, message);
```
We should use the new `removeMessagesFromGroup()` method - this one returns the whole group and then we discard it - ok for a simple store but not for a persistent store.

In fact, this method should be deprecated - this is the only place in the framework (aside from tests) that this method is called.

artembilan replied:
Will be changed. But we definitely use it here just only for the `SimpleMessageStore`.
Let's deprecate it in the separate JIRA!
Having such a big change here I even not sure that we will be able to back port anything from here...

garyrussell commented:
I am seeing a failure in `RedisChannelMessageStoreTests`.

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/aggregator/ResequencingMessageHandler.java:
```diff
@@ -81,13 +83,21 @@ protected void afterRelease(MessageGroup messageGroup, Collection<Message<?>> co
+				if (this.messageStore instanceof SimpleMessageStore
+						&& completedMessages.size() == messageGroup.size()) {
+					((SimpleMessageGroup) messageGroup).clear();
+					((SimpleMessageGroup) messageGroup).setLastModified(System.currentTimeMillis());
```
Unfortunately, this fix isn't good enough. We end up with the same `non-released permit` issue here again with limited `MessageStore`:
> org.springframework.messaging.MessagingException: SimpleMessageStore was out of capacity at for individual group, try constructing it with a larger capacity.

I still don't see way how to fix...

garyrussell replied:
Maybe add `clear()` to `SimpleMessageStore` and reset the upper bound?

artembilan replied:
To `SimpleMessageStore`...
Yes, that's perfect!
We can add it to the `MGS` contract already only in the `4.3`.

Thank you for the pointer! :smile:

garyrussell replied:
Maybe a sub-interface `ClearableMessageStore` in 4.2 and pull it up in 4.3.

artembilan replied:
Don't see reason. we need the "hard" clean only for the `SMS` with `UpperBound`.
All other cases we can consider as new feature and really can address with the new point release.

garyrussell replied:
ok

artembilan committed:
Add `SimpleMessageStore.clearMessageGroup()`

artembilan commented:
Pushed

garyrussell commented:
Some polish here: https://github.com/garyrussell/spring-integration/commit/0eec2b4331ac4da7b3a47cb9f708919374f6a623

I am still seeing failures in the redis tests:

    org.springframework.integration.redis.store.RedisChannelMessageStoreTests > testPriority FAILED
        java.lang.AssertionError at RedisChannelMessageStoreTests.java:130

artembilan commented:
This can be merged now.
Additional review is welcome!

garyrussell commented:
I don't see my polish applied, or addressing the redis test issue (which fails because we now use a Set instead of a Queue).

I thought we discussed that we somehow need to continue to support the existing behavior - or at least provide a hook of some kind to allow it.

artembilan commented:
Oh! Sorry. That was late evening...
Fixing now...

I'll come back with comments to other concerns here, when I refresh the task in my mind.

artembilan committed:
Accept polishing and fix `RedisChannelMessageStoreTests`

artembilan commented:
Pushed @garyrussell 's polishing and fix for the `RedisChannelMessageStoreTests`.

Some other work will be done later.

artembilan committed:
`@Deprecated` `MessageGroupStore.removeMessageFromGroup()`

Fix some typos

artembilan commented:
Pushed `@Deprecated` `MessageGroupStore.removeMessageFromGroup()` .

@garyrussell , since we are here in a fresh minor release, I think we can go ahead without:

> I thought we discussed that we somehow need to continue to support the existing behavior - or at least provide a hook of some kind to allow it

I can, of course, add `BlockingQueue<Message<?>> messages` back and some addition `queueBasedMessageGroup = true/false` option to the `MessageGroupStore`. But will it have so much sense, if we are fine with the `Set<Message<?>>`?
Anyway end-users should follow with our Migration Guide on upgrade.

Your thoughts?

garyrussell commented:
I agree it's a corner case; I don't think it's so easy to provide a hook because `new SimpleMessageGroup()` is used all over the place.

I am fine with the change, but I'd prefer to see all such code replaced with a factory (e.g. prototype bean in the context), to provide more flexibility.

artembilan commented:
> but I'd prefer to see all such code replaced with a factory (e.g. prototype bean in the context), to provide more flexibility.

E-e-e... Sorry, I don't understand. What do you mean? `FactoryBean` for the `SimpleMessageStore` ?
What that factory should do?
Thanks

garyrussell commented:
No; not the store, `SimpleMessageGroup` - that's the class with the behavior change.

artembilan committed:
Introduce `SimpleMessageGroupFactory`

artembilan commented:
Pushed.
There are some other places to consider to rework for that `SimpleMessageGroupFactory`, but I'd like to know if I go right way, first of all.

artembilan committed:
Extract `MessageGroupFactory` and address PR comments

artembilan commented:
Pushed `MessageGroupFactory` contract.
There is need some JavaDocs for new API and maybe some docs on the matter.
But let me know, first of all, that we are on the right way.
Thanks.

garyrussell commented:
I am ok with the current state of this PR if you want to continue with other changes. Or I can merge (if it builds) and you can start a new PR. Up to you.

artembilan commented:
Good. Thank you!
Let me rebase it, test locally and fix, if that.
I'll let you know here when it will be ready.

artembilan committed:
Polishing after rebase

artembilan commented:
Pushed after rebase and Copyright fix, and `//NOSONAR` for the `FileWritingMH`

garyrussell commented:
>There is need some JavaDocs for new API and maybe some docs on the matter.

So I should merge and that will be a new PR?

artembilan commented:
> So I should merge and that will be a new PR?

Doh... Yes, of course we need doc. They will be here.
Thank you for catching! :smile:

artembilan committed:
JavaDocs and Reference Manual

artembilan commented:
Pushed Docs.

artembilan committed:
Fix JavaDocs

artembilan commented:
Pushed fix for JavaDocs.

artembilan commented:
@garyrussell , would you mind to review the last couple commits to close this lo-o-ong track and 3 JIRAs to have green graphic in JIRA project again :smile:

garyrussell commented:
Rebased, squashed, merged as 201f0fc2b16c38ca7c7ac09a1f4edc581746a750

Congrats @artembilan !!

artembilan commented:
And especial thanks to @celandro , who initiated the work and has forced us to reconsider the stuff.

Looking forward for more similar valuable contributions! :smile:
